File: shared/models/CareRecipientProfile.js
/**
 * CareRecipientProfile — ESM/JS mirror of the TypeScript contract.
 *
 * This file is the backend-consumable mirror of
 * `./CareRecipientProfile.ts`. JSDoc `@typedef` blocks give the backend
 * editor-level type awareness without pulling TypeScript into the Node
 * build. The frozen constant objects below are the runtime enum values —
 * import them from route handlers and validators so every call site agrees
 * on the same strings the TS + Swift contracts use.
 *
 * Rule: if you change a field or value here, you MUST update the `.ts`
 * and the Swift model in the same change. Three files, one shape.
 */

// ─── Enum constants (runtime) ───────────────────────────────────────────────

/** Risk stratification surfaced on the profile header. */
export const RISK_LEVELS = Object.freeze({
  low: "low",
  moderate: "moderate",
  high: "high",
})

/** Supported data-source integrations. Must match `DataSourceType` in TS. */
export const DATA_SOURCE_TYPES = Object.freeze({
  appleHealth: "apple_health",
  epic: "epic",
  fitbit: "fitbit",
  garmin: "garmin",
  ring: "ring",
  fallDetection: "fall_detection",
})

/** Connection state the UI renders for each data source. */
export const DATA_SOURCE_STATUSES = Object.freeze({
  connected: "connected",
  notConnected: "not_connected",
  error: "error",
})

/** Roles a care-team member can hold. Must match `CareTeamRole` in TS. */
export const CARE_TEAM_PROFILE_ROLES = Object.freeze({
  primaryCaregiver: "primary_caregiver",
  familyMember: "family_member",
  clinician: "clinician",
  careCoordinator: "care_coordinator",
})

/** Access levels for care-team members. Must match `CareTeamPermission` in TS. */
export const CARE_TEAM_PROFILE_PERMISSIONS = Object.freeze({
  fullAccess: "full_access",
  limitedAccess: "limited_access",
  clinicalAccess: "clinical_access",
  viewOnly: "view_only",
})

// ─── Type aliases for editor tooling (JSDoc) ────────────────────────────────

/**
 * @typedef {"low" | "moderate" | "high"} RiskLevel
 *
 * @typedef {"apple_health" | "epic" | "fitbit" | "garmin" | "ring" | "fall_detection"} DataSourceType
 *
 * @typedef {"connected" | "not_connected" | "error"} DataSourceStatus
 *
 * @typedef {"primary_caregiver" | "family_member" | "clinician" | "care_coordinator"} CareTeamRole
 *
 * @typedef {"full_access" | "limited_access" | "clinical_access" | "view_only"} CareTeamPermission
 *
 * @typedef {Object} CareTeamMember
 * @property {string} id
 * @property {string} name
 * @property {CareTeamRole} role
 * @property {CareTeamPermission} permission
 *
 * @typedef {Object} CareRecipientContact
 * @property {string} [phone]
 * @property {string} [address]
 *
 * @typedef {Object} EmergencyContact
 * @property {string} name
 * @property {string} phone
 * @property {string} [relationship]
 *
 * @typedef {Object} CareTeam
 * @property {string} primaryCaregiver
 * @property {CareTeamMember[]} members
 *
 * @typedef {Object} HealthBackground
 * @property {string[]} conditions
 * @property {string[]} allergies
 * @property {string[]} medications
 * @property {string} [mobilityStatus]
 * @property {string} [fallRiskNotes]
 *
 * @typedef {Object} DataSource
 * @property {DataSourceType} type
 * @property {DataSourceStatus} status
 * @property {string} [lastSyncedAt]
 * @property {string} [errorMessage]
 *
 * @typedef {Object} BaselineRange
 * @property {number} min
 * @property {number} max
 *
 * @typedef {Object} Baseline
 * @property {BaselineRange} [steps]
 * @property {BaselineRange} [sleepHours]
 * @property {BaselineRange} [restingHeartRate]
 * @property {string} [bloodPressure]
 * @property {string} [lastUpdated]
 *
 * @typedef {Object} RecentNote
 * @property {string} id
 * @property {string} content
 * @property {string} author
 * @property {string} createdAt
 *
 * @typedef {Object} CareRecipientProfile
 * @property {string} id
 * @property {string} name
 * @property {number} age
 * @property {string} dateOfBirth
 * @property {string} [gender]
 * @property {string[]} primaryConditions
 * @property {RiskLevel} riskLevel
 * @property {CareRecipientContact} contact
 * @property {EmergencyContact} emergencyContact
 * @property {CareTeam} careTeam
 * @property {HealthBackground} healthBackground
 * @property {DataSource[]} dataSources
 * @property {Baseline} baseline
 * @property {RecentNote[]} recentNotes
 * @property {string} lastUpdated
 */

export {}

File: apps/ios/NARTHECare/Models/CareRecipientProfile.swift
import Foundation

/// Swift mirror of `shared/contracts/careRecipientProfile.schema.json`.
///
/// This file is the iOS side of the cross-platform data contract for
/// `GET /care-recipients/:id/profile`. It is hand-kept in lockstep with:
///
/// - `shared/contracts/careRecipientProfile.schema.json` (source of truth)
/// - `shared/models/CareRecipientProfile.ts` (web TS mirror)
/// - `shared/models/CareRecipientProfile.js` (backend JS mirror)
///
/// Field names are camelCase and match the JSON payload exactly, so no
/// `CodingKeys` remapping is needed. Enum raw values are snake_case to
/// match the schema's enum strings.
///
/// All structs are `Codable, Sendable` so SwiftUI views and view models
/// can hold them across `Task`/actor boundaries without warnings.
/// Never log instances of these structs — most fields are PHI.

// MARK: - Enums

enum RiskLevel: String, Codable, Sendable, CaseIterable {
  case low
  case moderate
  case high
}

enum DataSourceType: String, Codable, Sendable, CaseIterable {
  case appleHealth = "apple_health"
  case epic
  case fitbit
  case garmin
  case ring
  case fallDetection = "fall_detection"
}

enum DataSourceStatus: String, Codable, Sendable, CaseIterable {
  case connected
  case notConnected = "not_connected"
  case error
}

enum CareTeamRole: String, Codable, Sendable, CaseIterable {
  case primaryCaregiver = "primary_caregiver"
  case familyMember = "family_member"
  case clinician
  case careCoordinator = "care_coordinator"
}

enum CareTeamPermission: String, Codable, Sendable, CaseIterable {
  case fullAccess = "full_access"
  case limitedAccess = "limited_access"
  case clinicalAccess = "clinical_access"
  case viewOnly = "view_only"
}

// MARK: - Nested payload types

struct CareTeamMember: Codable, Sendable, Identifiable, Hashable {
  let id: String
  let name: String
  let role: CareTeamRole
  let permission: CareTeamPermission
}

struct CareRecipientContact: Codable, Sendable, Hashable {
  let phone: String?
  let address: String?
}

struct EmergencyContact: Codable, Sendable, Hashable {
  let name: String
  let phone: String
  let relationship: String?
}

struct CareTeam: Codable, Sendable, Hashable {
  let primaryCaregiver: String
  let members: [CareTeamMember]
}

struct HealthBackground: Codable, Sendable, Hashable {
  let conditions: [String]
  let allergies: [String]
  let medications: [String]
  let mobilityStatus: String?
  let fallRiskNotes: String?
}

struct DataSource: Codable, Sendable, Identifiable, Hashable {
  let type: DataSourceType
  let status: DataSourceStatus
  let lastSyncedAt: String?
  let errorMessage: String?

  /// `type` is unique within a profile (the backend never returns two
  /// rows for the same integration family) so it doubles as a stable
  /// id for SwiftUI `ForEach`.
  var id: DataSourceType { type }
}

struct BaselineRange: Codable, Sendable, Hashable {
  let min: Double
  let max: Double
}

struct Baseline: Codable, Sendable, Hashable {
  let steps: BaselineRange?
  let sleepHours: BaselineRange?
  let restingHeartRate: BaselineRange?
  let bloodPressure: String?
  let lastUpdated: String?
}

struct RecentNote: Codable, Sendable, Identifiable, Hashable {
  let id: String
  let content: String
  let author: String
  let createdAt: String
}

// MARK: - Top-level profile

struct CareRecipientProfile: Codable, Sendable, Identifiable, Hashable {
  let id: String
  let name: String
  let age: Int
  let dateOfBirth: String
  let gender: String?

  let primaryConditions: [String]
  let riskLevel: RiskLevel

  let contact: CareRecipientContact
  let emergencyContact: EmergencyContact
  let careTeam: CareTeam
  let healthBackground: HealthBackground
  let dataSources: [DataSource]
  let baseline: Baseline
  let recentNotes: [RecentNote]

  let lastUpdated: String
}

/// Envelope returned by `GET /care-recipients/:id/profile`.
///
/// Kept as its own type so we can add sibling fields (e.g. `warnings`)
/// without a breaking change to the top-level payload.
struct CareRecipientProfileResponse: Codable, Sendable {
  let careRecipient: CareRecipientProfile
}
